packages/core/data-transfer/src/strapi/remote/handlers/utils.ts

```diff
@@ -0,0 +1,297 @@
+   */
+  teardown(): Promise<void> | void;
+
+  /**
+   * Lifecycle called to cleanup the transfer state
+   */
+  cleanup(): Promise<void> | void;
+
+  // Transfer Commands
+  init(...args: unknown[]): unknown;
+  end(...args: unknown[]): unknown;
+  status(...args: unknown[]): unknown;
+
+  // Events
+  onMessage(message: RawData, isBinary: boolean): Promise<void> | void;
+  onClose(code: number, reason: Buffer): Promise<void> | void;
+  onError(err: Error): Promise<void> | void;
+}
+
+export const handlerFactory =
+  <T extends Partial<Handler>>(implementation: (proto: Handler) => T) =>
+  (options: IHandlerOptions) => {
+    const { verify, server: serverOptions } = options ?? {};
+
+    const wss = new WebSocket.Server({ ...serverOptions, noServer: true });
+
+    return async (ctx: Context) => {
+      handleWSUpgrade(wss, ctx, (ws) => {
+        const state: TransferState = { id: undefined };
+
+        const prototype: Handler = {
+          // Transfer ID
+          get transferID() {
+            return state.id;
+          },
+
+          set transferID(id) {
+            state.id = id;
+          },
+
+          // Started at
+          get startedAt() {
+            return state.startedAt;
+          },
+
+          set startedAt(timestamp) {
+            state.startedAt = timestamp;
+          },
+
+          isTransferStarted() {
+            return this.transferID !== undefined && this.startedAt !== undefined;
+          },
+
+          assertValidTransfer() {
+            const isStarted = this.isTransferStarted();
+
+            if (!isStarted) {
+              throw new Error('Invalid Transfer Process');
+            }
+          },
+
+          assertValidTransferCommand(command: TransferCommand) {
+            const isDefined = typeof this[command] === 'function';
+            const isValidTransferCommand = VALID_TRANSFER_COMMANDS.includes(command);
+
+            if (!isDefined || !isValidTransferCommand) {
+              throw new Error('Invalid transfer command');
+            }
+          },
+
+          respond(uuid, e, data) {
+            return new Promise<void>((resolve, reject) => {
+              if (!uuid && !e) {
+                reject(new Error('Missing uuid for this message'));
+                return;
+              }
+
+              const payload = {
+                uuid,
+                data: data ?? null,
+                error: e
+                  ? {
+                      code: e?.name ?? 'ERR',
+                      message: e?.message,
+                    }
+                  : null,
+              };
+
+              this.send(payload, (error) => (error ? reject(error) : resolve()));
+            });
+          },
+
+          send(message, cb) {
+            let payload;
+
+            try {
+              payload = JSON.stringify(message);
+            } catch {
+              payload = message;
+            }
+
+            ws.send(payload, cb);
```

Comment:
I can't build because there's a type error here. It looks like we just need to add a typeguard to make sure payload is a `BufferLike` -- which unfortunately doesn't appear to be exported by 'ws', but if we're only expecting payload to be a string checking that will be good enough to check

Reply:
This should be fixed by [5120ae4](https://github.com/strapi/strapi/pull/16025/commits/5120ae454d93a24e30fcf4aa1efa1b03a1ef2516)